# **02 - Exploratory Data Analysis**

## Objectives

- Load and inspect the cleaned Online Retail Transaction dataset.
- Explore overall revenue and sales performance.
- Analyse revenue and sales volume over time.
- Investigate product and customer purchasing patterns.
- Compare geographic performance across countries.
- Explore relevant distributions and relationships within the data.
- Identify key insights that can support further visualisation and machine learning analysis.

## Inputs

- Cleaned Online Retail Transaction dataset produced during the ETL process and stored in the `outputs` folder.

## Outputs

- Exploratory analysis of the Online Retail Transaction dataset.
- Key findings and visualisations relating to revenue, sales volume, customers, products and geographic performance.
- Insights to support subsequent stages of the project.

## Additional Comments

- Analysis will be exploratory, with findings and interpretations documented throughout the notebook.

---

# Change working directory

In [1]:
import os

current_dir = os.getcwd()
print("Current directory:", current_dir)

if os.path.basename(current_dir) == "jupyter_notebooks":
    os.chdir(os.path.dirname(current_dir))
elif os.path.basename(current_dir) == "vscode-projects":
    os.chdir("Online Retail Transaction Analysis")

current_dir = os.getcwd()
print("Working directory set to:", current_dir)

Current directory: /Users/sahraosman/Documents/vscode-projects/Online Retail Transaction Analysis/jupyter_notebooks
Working directory set to: /Users/sahraosman/Documents/vscode-projects/Online Retail Transaction Analysis


---

# Section 1 -  Import libraries

Here I will be importing the libraries that I will be using in this notebook.

In [2]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns


---

# Section 2 - Load the dataset

Here I'll be loading the cleaned Online Retail Transaction dataset that was produced during the ETL process.

In [3]:
df = pd.read_csv("outputs/Online_Retail_Cleaned.csv")

/var/folders/cx/w0ydrn213c14xglf6wmp_2l00000gn/T/ipykernel_73141/3308739013.py:1: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("outputs/Online_Retail_Cleaned.csv")


In [4]:
df.dtypes

InvoiceNo       object
StockCode       object
Description     object
Quantity         int64
InvoiceDate     object
UnitPrice      float64
CustomerID       int64
Country         object
Revenue        float64
dtype: object

The initial import produced a mixed data type warning for `InvoiceNo`. Inspection of the imported data types also showed that `InvoiceDate` had been loaded as an object rather than a datetime.

In [5]:
df["InvoiceNo"].apply(type).value_counts()

InvoiceNo
<class 'int'>    459342
<class 'str'>     65536
Name: count, dtype: int64

Further inspection showed that `InvoiceNo` contained both integer and string values. Since invoice numbers are identifiers rather than numerical measures, the column will be explicitly loaded as a string. `InvoiceDate` will also be parsed as a datetime to support time-based analysis.

In [ ]:
df = pd.read_csv(
    "outputs/Online_Retail_Cleaned.csv",
    dtype={"InvoiceNo": str},
    parse_dates=["InvoiceDate"],
)

In [7]:
df.dtypes

InvoiceNo              object
StockCode              object
Description            object
Quantity                int64
InvoiceDate    datetime64[ns]
UnitPrice             float64
CustomerID              int64
Country                object
Revenue               float64
dtype: object

In [8]:
df["InvoiceNo"].apply(type).value_counts()

InvoiceNo
<class 'str'>    524878
Name: count, dtype: int64

The corrected import confirms that all `InvoiceNo` values are now consistently loaded as strings and `InvoiceDate` has been successfully parsed as a datetime. The cleaned dataset is now loaded with appropriate data types for exploratory analysis.

In [9]:
df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,Revenue
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850,United Kingdom,15.30
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850,United Kingdom,20.34
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850,United Kingdom,22.00
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850,United Kingdom,20.34
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850,United Kingdom,20.34


In [ ]:
print("Number of rows: ", df.shape[0])

Number of rows:  524878


In [ ]:
print("Number of columns: ", df.shape[1])

Number of columns:  9


The dataset contains 524,878 records and 9 columns, matching the final dataset produced during the ETL process. An initial preview also confirms that the cleaned data and engineered `Revenue` feature have loaded successfully and are ready for exploratory analysis.